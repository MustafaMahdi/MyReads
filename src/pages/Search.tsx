import { useState } from 'react';
import { Link } from 'react-router-dom';
import { ShelvesType } from '../defines/defines';
import { Book as BookInterface } from "../models/book";
import Book from '../components/Book';
import * as BooksAPI from "../BooksAPI";



const Search: React.FC<{
    books: BookInterface[],
    changeShelfHandler: (book: BookInterface, selectedShelf: string) => void
}> = (props) => {
    const [searchInput, setSearchInput] = useState<string>('');
    const [searchBooksResult, setSearchBooksResult] = useState<BookInterface[]>([]);

    const getSearchResult = (searchQuery:string) => {

        if (searchQuery.length !== 0) {
            BooksAPI.search(searchQuery).then((matchedBooks) => {
                if (matchedBooks.error) {
                    setSearchBooksResult([]);
                } else {
                    setSearchBooksResult(matchedBooks);
                }
            });
        } else {
            setSearchBooksResult([]);
        }
    }
    const searchInputHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
        setSearchInput(event.target.value);
        getSearchResult(event.target.value);
    }

    const matchedBooks = searchBooksResult.map(matchedBook => {
        let shelf = ShelvesType.None;
        props.books.forEach(book => {
            if (book.id === matchedBook.id) {
                shelf = book.shelf
            }
        })
        return (
            <li key={matchedBook.id}>
                <Book
                    book={matchedBook}
                    onChangeShelf={props.changeShelfHandler}
                    currentShelf={shelf}
                />
            </li>

        )
    });

    return (
        <div className="search-books">
            <div className="search-books-bar">
                <Link to="/" className="close-search" >Close</Link>
                <div className="search-books-input-wrapper">
                    <input
                        type="text"
                        placeholder="Search by Title Or Author"
                        value={searchInput}
                        onChange={searchInputHandler}
                    />
                </div>
            </div>

            <div className="search-books-results">
                <ol className="books-grid">
                    {
                        matchedBooks
                    }
                </ol>
            </div>
        </div>
    );
};

export default Search;