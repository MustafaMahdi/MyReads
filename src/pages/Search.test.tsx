import React from 'react';
import { render, screen } from '@testing-library/react';
import Search from './Search';
import { BrowserRouter, Link, Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { createMemoryHistory } from 'history';
import { ShelvesType } from '../defines/defines';
import { Book as BookInterface } from "../models/book";
import Book from '../components/Book';
import userEvent from '@testing-library/user-event';



describe('Search component', () => {
  const props = {
    books: [{
      authors: ['Thomas HelwysRichard'],
      title: 'A Short Declaration of the Mystery of Iniquity',
      shelf: ShelvesType.None, id: "id",
      imageLinks: {
        thumbnail: "http://books.google.com/books/content?id=sJf1vQAACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api"
      }
    }],
    changeShelfHandler: (book: BookInterface, selectedShelf: string) => { },
  };
  it('renders Search Component', () => {
    render(<BrowserRouter><Search {...props} /></BrowserRouter>);
  });

  it('should route to BookShelves page when click on the link action', () => {
    const component = renderer.create(
      <BrowserRouter><Link to="/" /></BrowserRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('current Route should be /bookShelves', () => {
    const history = createMemoryHistory({ initialEntries: ['/search'] });
    const { getByText } = render(
      <Router history={history}>
        <Search {...props} />
      </Router>
    );
    expect(history.location.pathname).toBe('/search');
  });

  it('should render Book Component inside Search component by check the title of book which rendered at the Book component', () => {
    render(<BrowserRouter><Search {...props} /></BrowserRouter>);
    render(<Book book={props.books[0]} currentShelf={ShelvesType.CurrentlyReading} onChangeShelf={props.changeShelfHandler} />);
    expect(screen.getByText(/A Short Declaration of the Mystery of Iniquity/i)).toBeVisible();
  });

  it('test search input', () => {
    render(<BrowserRouter><Search {...props} /></BrowserRouter>);

    userEvent.type(screen.getByRole('textbox'), 'Thomas HelwysRichard')
    expect(screen.getByRole('textbox')).toHaveValue('Thomas HelwysRichard')
  })
});

