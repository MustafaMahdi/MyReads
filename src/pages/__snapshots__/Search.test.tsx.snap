// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should route to BookShelves page when click on the link action 1`] = `
<a
  href="/"
  onClick={[Function]}
/>
`;
